src/libcore/num/mod.rs

```diff
@@ -1471,3 +1472,73 @@ impl fmt::Display for ParseIntError {
 }
 
 pub use num::dec2flt::ParseFloatError;
+
+
+// Conversion traits for primitive integer types
+// Conversions T -> T are covered by a blanket impl and therefore excluded
+macro_rules! impl_from {
+    ($Small: ty, $Large: ty) => {
+        impl From<$Small> for $Large {
+            #[stable(feature = "lossless_int_conv", since = "1.5.0")]
+            #[inline]
+            fn from(small: $Small) -> $Large {
+                small as $Large
+            }
+        }
+    }
+}
+
+// Unsigned -> Unsigned
+impl_from! { u8, u16 }
+impl_from! { u8, u32 }
+impl_from! { u8, u64 }
+impl_from! { u8, usize }
+impl_from! { u16, u32 }
+impl_from! { u16, u64 }
+impl_from! { u16, usize }
+impl_from! { u32, u64 }
+#[cfg(any(target_pointer_width = "64", target_pointer_width = "32"))]
+impl_from! { u32, usize }
+#[cfg(target_pointer_width = "64")]
```

Comment:
@petrochenkov I'm not sure if you're joking, but no one is asking you to "surrender it just like that". We're trying to tease out the tradeoffs here, and as @alexcrichton says this is a significant departure from how we've handled platform specificity in the past. I've cited a downside here, that this makes it very easy to accidentally introduce implicit arch dependencies. I would like to explore the tradeoff space here and try to come to some consensus.

In the past when we've discussed this issue, we'd arrived at a compromise position:

- Allow implicitly going from 32 bit to `usize` on both 32- and 64-bit arches.
- Do not allow implicitly going from 64 bit to `usize` on any arch.

This biases us in the direction of compatibility across 32- and 64-bit systems, while covering what I expect is the common case of going from 32 bit to pointer size. And OTOH, if you try with a 64 bit value, you will be alerted to the fact that you'd be losing compat with 32 bit systems.

On the other hand, it means that compatibility with 16 bit systems is de-emphasized, making porting there more difficult. But that seems par for the course, and is likely to be difficult regardless of the decision here.